### ISSUE TITLE
Bug Report: Inconsistent Mypy Return Type Checking Across Identical Python/Mypy Environments Due to Build/Dependency Variations

### ISSUE THREAD

**Issue by WSYAH** [python/mypy#19680]
### Problem Description​

The same codebase produces ​different type-checking results​ under ​Python 3.9.21​ (detects an error) vs ​Python 3.9.23​ (no error), using the ​same Mypy version (1.6.0 and 1.17.1, tested both)​​ and ​identical Mypy configuration.

This inconsistency breaks team collaboration, as developers using different minor Python 3.9.x versions (e.g., 3.9.21 vs 3.9.23) get conflicting type-check results for the same code.

### Reproduction Steps​

​ Environment 1 (Error Detected)​:

> Python Version: 3.9.21
Mypy Command: mypy .
Mypy Config (pyproject.toml):
```
[tool.mypy]
ignore_missing_imports = true 
disallow_untyped_defs = true
explicit_package_bases = true
mypy_path = './'             
exclude = [
    '^script/.*',
    '.*setup.py$',
]
```

​Environment 2 (No Error)​:

> Python Version: 3.9.23
Mypy Command: mypy .
Same Mypy Config and Version (1.6.0 / 1.17.1).
Result: ​No errors reported​ for the same file/code.

Here is the Mypy check results screenshot for two different Python versions
<img width="1481" height="223" alt="Image" src="https://github.com/user-attachments/assets/775980d0-45d2-4375-a0fb-35db85877858" />

**Comment by sterliakov** [comment]
I suspect this is not about python versions themselves - could you `pip freeze` in both environments and look for any differences? I'm almost certain some package versions are not the same, and third-party typing has changed between these releases.

**Comment by WSYAH** [comment]
> I suspect this is not about python versions themselves - could you `pip freeze` in both environments and look for any differences? I'm almost certain some package versions are not the same, and third-party typing has changed between these releases.

Sure, I will run pip freeze in both environments, and then ​list the differences between the third-party packages in a table format.


package | test | uff | isconsistent 
-- | -- | -- | --
asttokens | 3.0.0 | 3.0.0 | ✅
attrs | 25.3.0 | 25.3.0 | ✅
autocommand | 2.2.2 | - | ❌
backcall | 0.2.0 | 0.2.0 | ✅
backports.tarfile | 1.2.0 | - | ❌
beautifulsoup4 | 4.13.4 | 4.13.4 | ✅
black | 25.1.0 | 25.1.0 | ✅
bleach | 6.2.0 | 6.2.0 | ✅
certifi | 2025.8.3 | 2025.4.26 | ❌
charset-normalizer | 3.4.3 | 3.4.2 | ❌
click | 8.2.2 | 8.1.8 | ❌
comm | 0.2.3 | 0.2.2 | ❌
debugpy | 1.8.14 | 1.8.14 | ✅
decorator | 5.2.1 | 5.2.1 | ✅
defusedxml | 0.8.0rc2 | 0.7.1 | ❌
exceptiongroup | - | 1.3.0 | ❌
executing | 2.2.0 | 2.2.0 | ✅
fastjsonschema | 2.21.2 | 2.21.1 | ❌
flake8 | 7.2.0 | 7.2.0 | ✅
gitdb | 4.0.12 | 4.0.12 | ✅
GitPython | 3.1.41 | 3.1.41 | ✅
idna | 3.10 | 3.10 | ✅
importlib_metadata | 8.0.0 | 8.7.0 | ❌
iniconfig | - | 2.1.0 | ❌
inflect | 7.3.1 | - | ❌
ipykernel | 6.29.5 | 6.29.5 | ✅
ipynbname | 2025.8.0.0 | 2024.1.0.0 | ❌
ipython | 8.12.0 | 8.12.0 | ✅
jedi | 0.19.2 | 0.19.2 | ✅
Jinja2 | 3.1.6 | 3.1.6 | ✅
jsonschema | 4.25.1 | 4.24.0 | ❌
jsonschema-specifications | 2025.4.1 | 2025.4.1 | ✅
jupyter_client | 8.6.3 | 8.6.3 | ✅
jupyter_core | 5.8.1 | 5.8.1 | ✅
jupyterlab_pygments | 0.3.0 | 0.3.0 | ✅
MarkupSafe | 3.0.2 | 3.0.2 | ✅
matplotlib-inline | 0.1.7 | 0.1.7 | ✅
mccabe | 0.7.0 | 0.7.0 | ✅
mistune | 3.1.3 | 3.1.3 | ✅
more-itertools | 10.3.0 | - | ❌
mypy | 1.16.0 | 1.16.0 | ✅
mypy-protobuf | 3.4.0 | 3.4.0 | ✅
mypy_extensions | 1.1.0 | 1.1.0 | ✅
nbclient | 0.10.2 | 0.10.2 | ✅
nbconvert | 7.8.0 | 7.8.0 | ✅
nbformat | 5.9.2 | 5.9.2 | ✅
nest-asyncio | 1.6.0 | 1.6.0 | ✅
networkx | 3.0 | 3.0 | ✅
numpy | 1.26.4 | 1.26.4 | ✅
packaging | 25.0 | 25.0 | ✅
pandas | 1.4.3 | 1.4.3 | ✅
pandocfilters | 1.5.1 | 1.5.1 | ✅
parso | 0.8.4 | 0.8.4 | ✅
pathspec | 0.12.1 | 0.12.1 | ✅
pexpect | 4.9.0 | 4.9.0 | ✅
pickleshare | 0.7.5 | 0.7.5 | ✅
platformdirs | 4.2.2 | 4.3.8 | ❌
pluggy | - | 1.6.0 | ❌
prompt_toolkit | 3.0.51 | 3.0.51 | ✅
protobuf | 4.22.1 | 4.22.1 | ✅
psutil | 7.0.0 | 7.0.0 | ✅
ptyprocess | 0.7.0 | 0.7.0 | ✅
pure_eval | 0.2.3 | 0.2.3 | ✅
pycodestyle | 2.13.0 | 2.13.0 | ✅
pyDes | 2.0.1 | 2.0.1 | ✅
pyflakes | 3.3.2 | 3.3.2 | ✅
Pygments | 2.19.2 | 2.19.1 | ❌
pypinyin | 0.54.0 | 0.54.0 | ✅
python-dateutil | 2.9.0.post0 | 2.9.0.post0 | ✅
pytz | 2025.2 | 2025.2 | ✅
PyYAML | 6.0.2 | 6.0.2 | ✅
pyzmq | 27.0.1 | 26.4.0 | ❌
referencing | 0.36.2 | 0.36.2 | ✅
requests | 2.32.3 | 2.32.3 | ✅
rpds-py | 0.27.0 | 0.25.1 | ❌
six | 1.17.0 | 1.17.0 | ✅
smmap | 5.0.2 | 5.0.2 | ✅
soupsieve | 2.7 | 2.7 | ✅
stack-data | 0.6.3 | 0.6.3 | ✅
tinycss2 | 1.4.0 | 1.4.0 | ✅
tomli | 2.2.1 | 2.2.1 | ✅
tornado | 6.5.1 | 6.5.1 | ✅
traitlets | 5.14.3 | 5.14.3 | ✅
typeguard | 4.3.0 | - | ❌
types-protobuf | 6.30.2.20250516 | 6.30.2.20250516 | ✅
typing_extensions | 4.13.2 | 4.13.2 | ✅
uff | - | - (local package) | -
urllib3 | 2.4.0 | 2.4.0 | ✅
wcwidth | 0.2.13 | 0.2.13 | ✅
webencodings | 0.5.1 | 0.5.1 | ✅
zipp | 3.19.2 | 3.22.0 | ❌


**PS： I also installed Python 3.9.21 via conda in the test environment (to align the Python versions), but nothing changed. So it really isn't a Python version issue — you were right. So then I need to update the issue title.**

<img width="1315" height="151" alt="Image" src="https://github.com/user-attachments/assets/1b9c50ca-7096-43ea-bd2c-9e5c3ef05480" />

**the results of running command pip freeze:**
**Environment test:**
`asttokens==3.0.0
attrs==25.3.0
autocommand==2.2.2
backcall==0.2.0
backports.tarfile==1.2.0
beautifulsoup4==4.13.4
black==25.1.0
bleach==6.2.0
certifi==2025.8.3
charset-normalizer==3.4.3
click==8.2.2
comm==0.2.3
debugpy==1.8.14
decorator==5.2.1
defusedxml==0.8.0rc2
executing==2.2.0
fastjsonschema==2.21.2
flake8==7.2.0
gitdb==4.0.12
GitPython==3.1.41
idna==3.10
importlib_metadata==8.0.0
inflect==7.3.1
ipykernel==6.29.5
ipynbname==2025.8.0.0
ipython==8.12.0
jaraco.collections==5.1.0
jaraco.context==5.3.0
jaraco.functools==4.0.1
jaraco.text==3.12.1
jedi==0.19.2
Jinja2==3.1.6
jsonschema==4.25.1
jsonschema-specifications==2025.4.1
jupyter_client==8.6.3
jupyter_core==5.8.1
jupyterlab_pygments==0.3.0
MarkupSafe==3.0.2
matplotlib-inline==0.1.7
mccabe==0.7.0
mistune==3.1.3
more-itertools==10.3.0
mypy==1.16.0
mypy-protobuf==3.4.0
mypy_extensions==1.1.0
nbclient==0.10.2
nbconvert==7.8.0
nbformat==5.9.2
nest-asyncio==1.6.0
networkx==3.0
numpy==1.26.4
packaging==25.0
pandas==1.4.3
pandocfilters==1.5.1
parso==0.8.4
pathspec==0.12.1
pexpect==4.9.0
pickleshare==0.7.5
platformdirs==4.2.2
prompt_toolkit==3.0.51
protobuf==4.22.1
psutil==7.0.0
ptyprocess==0.7.0
pure_eval==0.2.3
pycodestyle==2.13.0
pyDes==2.0.1
pyflakes==3.3.2
Pygments==2.19.2
pypinyin==0.54.0
python-dateutil==2.9.0.post0
pytz==2025.2
PyYAML==6.0.2
pyzmq==27.0.1
referencing==0.36.2
requests==2.32.3
rpds-py==0.27.0
six==1.17.0
smmap==5.0.2
soupsieve==2.7
stack-data==0.6.3
tinycss2==1.4.0
tomli==2.2.1
tornado==6.5.1
traitlets==5.14.3
typeguard==4.3.0
types-protobuf==6.30.2.20250516
typing_extensions==4.13.2
-e git+ssh://git@git.woa.com/feature-store/ads-rfc.git@b450637c599826dd5a9f84a80757b231e02a41b0#egg=uff&subdirectory=feature_generation/rfc/uff/python
urllib3==2.4.0
wcwidth==0.2.13
webencodings==0.5.1
zipp==3.19.2
`
**Environment uff:**
`asttokens==3.0.0
attrs==25.3.0
backcall==0.2.0
beautifulsoup4==4.13.4
black==25.1.0
bleach==6.2.0
certifi==2025.4.26
charset-normalizer==3.4.2
click==8.1.8
comm==0.2.2
debugpy==1.8.14
decorator==5.2.1
defusedxml==0.7.1
exceptiongroup==1.3.0
executing==2.2.0
fastjsonschema==2.21.1
flake8==7.2.0
gitdb==4.0.12
GitPython==3.1.41
idna==3.10
importlib_metadata==8.7.0
iniconfig==2.1.0
ipykernel==6.29.5
ipynbname==2024.1.0.0
ipython==8.12.0
jedi==0.19.2
Jinja2==3.1.6
jsonschema==4.24.0
jsonschema-specifications==2025.4.1
jupyter_client==8.6.3
jupyter_core==5.8.1
jupyterlab_pygments==0.3.0
MarkupSafe==3.0.2
matplotlib-inline==0.1.7
mccabe==0.7.0
mistune==3.1.3
mypy==1.16.0
mypy-protobuf==3.4.0
mypy_extensions==1.1.0
nbclient==0.10.2
nbconvert==7.8.0
nbformat==5.9.2
nest-asyncio==1.6.0
networkx==3.0
numpy==1.26.4
packaging==25.0
pandas==1.4.3
pandocfilters==1.5.1
parso==0.8.4
pathspec==0.12.1
pexpect==4.9.0
pickleshare==0.7.5
platformdirs==4.3.8
pluggy==1.6.0
prompt_toolkit==3.0.51
protobuf==4.22.1
psutil==7.0.0
ptyprocess==0.7.0
pure_eval==0.2.3
py4j==0.10.9.5
pycodestyle==2.13.0
pyDes==2.0.1
pyflakes==3.3.2
Pygments==2.19.1
pypinyin==0.54.0
pyspark==3.3.2
pytest==8.4.1
python-dateutil==2.9.0.post0
pytz==2025.2
PyYAML==6.0.2
pyzmq==26.4.0
referencing==0.36.2
requests==2.32.3
rpds-py==0.25.1
six==1.17.0
smmap==5.0.2
soupsieve==2.7
stack-data==0.6.3
tinycss2==1.4.0
tomli==2.2.1
tornado==6.5.1
traitlets==5.14.3
types-protobuf==6.30.2.20250516
typing_extensions==4.13.2
-e git+ssh://git@git.woa.com/feature-store/ads-rfc.git@b450637c599826dd5a9f84a80757b231e02a41b0#egg=uff&subdirectory=feature_generation/rfc/uff/python
urllib3==2.4.0
wcwidth==0.2.13
webencodings==0.5.1
zipp==3.22.0
`

**Comment by WSYAH** [comment]
When building the environment using setup.py, if you first run `python setup.py develop`, certain type checking issues may not be detected — for example, as seen in the above **test** environment.

The root cause was found to be that both` python setup.py sdist `and `python setup.py develop `unexpectedly depend on the jaraco.functools package (triggering a ModuleNotFoundError: No module named 'jaraco.functools'), which should not normally be required. This revealed the underlying issue.

On the other hand, if you first set up the environment using `pip install -e . `(as in the above **uff** environment), the type checking works correctly and the issue is properly detected. The exact reason for this difference is unclear.
